libnavui-voice/src/main/java/com/mapbox/navigation/ui/voice/api/VoiceInstructionsDownloadTrigger.kt

```diff
@@ -0,0 +1,151 @@
+        TimeBasedNextVoiceInstructionsProvider(observableTime),
+        Time.SystemImpl
+    )
+
+    private val ignoredRouteUpdateReasons = setOf(
+        RoutesExtra.ROUTES_UPDATE_REASON_CLEAN_UP,
+        RoutesExtra.ROUTES_UPDATE_REASON_ALTERNATIVE,
+        RoutesExtra.ROUTES_UPDATE_REASON_REFRESH,
+    )
+    private var lastDownloadTime: Long = 0
+
+    /**
+     * See [RoutesObserver.onRoutesChanged].
+     */
+    override fun onRoutesChanged(result: RoutesUpdatedResult) {
+        if (result.reason in ignoredRouteUpdateReasons) {
+            return
+        }
+        result.navigationRoutes.firstOrNull()?.let {
+            val currentStep = it.directionsRoute.legs()?.firstOrNull()?.steps()?.firstOrNull()
+            if (currentStep != null) {
+                val progress = RouteProgressData(
+                    it.directionsRoute,
+                    0,
+                    0,
+                    currentStep.duration(),
+                    currentStep.distance()
+                )
+                triggerDownload(progress)
+            }
+        }
+    }
+
+    /**
+     * See [RouteProgressObserver.onRouteProgressChanged].
+     */
+    override fun onRouteProgressChanged(routeProgress: RouteProgress) {
+        if (shouldDownloadBasedOnTime()) {
+            val legIndex = routeProgress.currentLegProgress?.legIndex
+            val currentStepProgress = routeProgress.currentLegProgress?.currentStepProgress
+            val stepIndex = currentStepProgress?.stepIndex
+            val durationRemaining = currentStepProgress?.durationRemaining
+            val distanceRemaining = currentStepProgress?.distanceRemaining
+            ifNonNull(
+                legIndex,
+                stepIndex,
+                durationRemaining,
+                distanceRemaining
+            ) { legIndex, stepIndex, durationRemaining, distanceRemaining ->
+                val progressData = RouteProgressData(
+                    routeProgress.route,
+                    legIndex,
+                    stepIndex,
+                    durationRemaining,
+                    distanceRemaining.toDouble()
+                )
+                triggerDownload(progressData)
+            }
+        }
+    }
+
+    /**
+     * The method stops all work related to pre-downloading voice instructions and unregisters
+     * all related callbacks. It should be invoked from `Activity#onDestroy`.
+     */
+    fun destroy() {
+        speechApi.destroy()
+    }
+
+    private fun shouldDownloadBasedOnTime(): Boolean {
+        return timeProvider.seconds() >=
+            lastDownloadTime + observableTime * timePercentageToTriggerAfter
+    }
+
+    private fun triggerDownload(progressData: RouteProgressData) {
+        lastDownloadTime = timeProvider.seconds()
+        val nextInstructionsToDownload = nextVoiceInstructionsProvider
+            .getNextVoiceInstructions(progressData)
+        speechApi.predownload(nextInstructionsToDownload)
```

Comment:
We had this topic during the first iteration, but I still haven't got how it work. Sorry, I'm asking it again 🙂 
How are the predownloaded instructions cleanded up if user interrupts the ride before downloaded instructions are played?

For example:
1. User set a different route
2. Reroute happened
3. OS killed the navigation application

Reply:
TLDR: basically they won't be cleaned. Just the way the played ones won't be cleaned.
We only clean the files that we create ourselves. However, we don't clean any caches on TileStore side. And we can't. Capturing from a Slack thread:
> Resource cache is not cleared explicitly. TileStore has a unified quota for tiles and resources and will evict individual items when the quota is exceeded. Resources (including tiles) with the closest expiration date are removed until the quota is not exceeded anymore.

Reply:
> Just the way the played ones won't be cleaned.

I thought [this](https://github.com/mapbox/mapbox-navigation-android/blob/ca5276ba448a4002d7b771a23af1d9a2addd7b97/libnavui-voice/src/main/java/com/mapbox/navigation/ui/voice/api/MapboxVoiceApi.kt#L38) cleans played voice instructions, doesn't it?

Reply:
It only cleans the files created here: https://github.com/mapbox/mapbox-navigation-android/blob/ca5276ba448a4002d7b771a23af1d9a2addd7b97/libnavui-voice/src/main/java/com/mapbox/navigation/ui/voice/api/MapboxVoiceApi.kt#L27
Predownloading does not create this file.

Reply:
Hm, interesting 🤔 
What if we have a long route, with predictive cache controller enabled (downloads tiles for the whole route ahead), and a tile store quota which is not enough to fit voice instructions and tiles for the whole route? Can downloaded voice instructions be evicted in favour of tiles before they are played? How do you think, is that scenario possible?

Reply:
> Can downloaded voice instructions be evicted in favour of tiles before they are played?

Right now not because they are being downloaded on-demand. After this PR - well, I think that is theoretically possible. I know that they have cache clearing mechanism planned but it's not available yet and I don't think that this is expected very soon.
Thanks for your observation, I think it's important so I'll need to understand it better. I'll talk to the Core SDK team again and run some tests.

Reply:
You're right. I was able to reproduce it in an easier way: if I set a small quota (`TileStore#setOptions`), some voice instructions may be removed because of the new ones, although the old ones have not yet been played.

Reply:
The other problem is that we won't know when the resource was removed from file cache by Core SDK so we might think it's downloaded although it's not. 
I'll try to think of a way to handle it.